src/topp/GNPSExport.cpp

```diff
@@ -79,297 +147,368 @@ using namespace std;
-
-            feature_stream << "MSLEVEL=2" << endl;
-            feature_stream << "CHARGE=" << to_string(charge == 0 ? 1 : charge) << "+" << endl;
-            feature_stream << "PEPMASS=" << peptide.first.second.getMZ() << endl;
-            feature_stream << "FILE_INDEX=" << peptide.second.second << endl;
-            feature_stream << "RTINSECOND=" << peptide.first.second.getRT() << endl;
-
-            auto ms2_scan = ms_maps[peptide.second.first][peptide.second.second];
-            // sort spectra
-            sort (ms2_scan.begin(), ms2_scan.end(), [](const Peak1D& a, const Peak1D& b)
-            {
-              return a.getMZ() > b.getMZ();
-            });
-            // ms2_scan.sortByIntensity(true);
-
-            for (Size l = 0; l < ms2_scan.size(); l++)
-            {
-              feature_stream << ms2_scan[l].getMZ() << "\t" << (int) ms2_scan[l].getIntensity() << endl;
-            }
-
-            feature_stream << "END IONS" << endl << endl;
-          }
-          feature_count++;
-        }
-        else // merged spectra
+        for (pair<int,int> pept : pepts)
         {
-          // map mz to intensity
-          map<double,int> ms2_block;
-
-          // MSExperiment exp;
-
-          const BinnedSpectrum binned_highest_int(ms_maps[peptides[0].second.first][peptides[0].second.second], BinnedSpectrum::DEFAULT_BIN_WIDTH_HIRES, false, 1, BinnedSpectrum::DEFAULT_BIN_OFFSET_HIRES);
-
-          for (const auto& peptide : peptides)
-          {
-            int map_index = peptide.second.first;
-            int spectra_index = peptide.second.second;
-            // int highest_binned_intensity = peptide.first.first;
-            // auto highest_peptide_identification = peptide.first.second;
+          int map_index = pept.first;
+          int spec_index = pept.second;
 
-            auto spectrum = ms_maps[map_index][spectra_index];
+          MSExperiment& specs = getSpectraAtIndex(mzml_file_paths, specs_list, map_index);
 
-            const BinnedSpectrum binned_spectrum(spectrum, BinnedSpectrum::DEFAULT_BIN_WIDTH_HIRES, false, 1, BinnedSpectrum::DEFAULT_BIN_OFFSET_HIRES);
+          auto& test_spec = specs[spec_index];
+          const BinnedSpectrum binned_spectrum(test_spec, BinnedSpectrum::DEFAULT_BIN_WIDTH_HIRES, false, 1, BinnedSpectrum::DEFAULT_BIN_OFFSET_HIRES);
 
-            BinnedSpectralContrastAngle bsca;
-            double cosine_sim = bsca(binned_highest_int, binned_spectrum);
-            // OPENMS_LOG_DEBUG << cosine_sim << " >= " << cos_sim << endl;
+          BinnedSpectralContrastAngle bsca;
+          double cos_sim = bsca(binned_highest_int, binned_spectrum);
 
-            // compare calculated cosine sim to binned highest int
-            if (cosine_sim >= cos_sim)
+          if (cos_sim >= cos_sim_threshold)
+          {
+            for (auto spec_iter = test_spec.begin(); spec_iter < test_spec.end(); spec_iter++)
```

Comment:
again: no need for this conversion. Just pass a spectrum to `generateMSMSSpectrumBins`

Reply:
I am attempting to merge multiple MSSpectrum scans. I browsed the namespace and was unable to find a way to append/extend a MSSpectrum with more scans. What would you suggest to be most optimal?